### ISSUE TITLE
whitespace: failed to get line path/parse/yaccpar:329

### ISSUE THREAD

**Issue by theory** [golangci/golangci-lint#3967]
### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've read the typecheck section of the FAQ (https://golangci-lint.run/usage/faq/#why-do-you-have-typecheck-errors).
- [X] Yes, I've tried with the standalone linter if available (e.g., gocritic, go vet, etc.). (https://golangci-lint.run/usage/linters/)

### Description of the problem

Similar to #2788, but chokes on files generated by goyacc, it chokes on lines like this:

``` go
//line yaccpar:1
```

Oddly, if I remove lines starting with `//line yacc`, I get a different error, where it looks like `golangci-lint` tries to validate the `.y` grammar file? The error is:

```
can't run linter goanalysis_metalinter: whitespace: failed to get line goyacc-tutorial/calculator/grammar_03_final.y:462: invalid file line index0 (461) >= len(fc) (59)
```

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version 1.53.3 built with go1.20.5 from 2dcd82f on 2023-06-14T21:13:17Z
```

</details>

### Configuration

<details>

```yaml
linters:
  enable-all: true
  disable:
    - funlen
    - varnamelen
    - cyclop
    - exhaustruct
    - testpackage
    - gomnd # Failed to get ignored-functions to work below
    # Disable deprecated checks
    - exhaustivestruct
    - varcheck
    - maligned
    - structcheck
    - ifshort
    - golint
    - deadcode
    - interfacer
    - scopelint
    - nosnakecase

issues:
  fix: true
```

</details>

### Go environment

<details>

```console
$ go version && go env
go version go1.20.3 darwin/arm64
GO111MODULE=""
GOARCH="arm64"
GOBIN=""
GOCACHE="/Users/david/Library/Caches/go-build"
GOENV="/Users/david/Library/Application Support/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="arm64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/david/.gvm/pkgsets/go1.20.3/global/pkg/mod"
GONOPROXY="github.com/nytm/*,github.com/NYTimes/*,github.com/nytimes/*"
GONOSUMDB="github.com/nytm/*,github.com/NYTimes/*,github.com/nytimes/*"
GOOS="darwin"
GOPATH="/Users/david/.gvm/pkgsets/go1.20.3/global"
GOPRIVATE="github.com/nytm/*,github.com/NYTimes/*,github.com/nytimes/*"
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/Users/david/.gvm/gos/go1.20.3"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/Users/david/.gvm/gos/go1.20.3/pkg/tool/darwin_arm64"
GOVCS=""
GOVERSION="go1.20.3"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/david/dev/go/goyacc-tutorial/go.mod"
GOWORK=""
CGO_CFLAGS="-O2 -g"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-O2 -g"
CGO_FFLAGS="-O2 -g"
CGO_LDFLAGS="-O2 -g"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/fl/mkyy75ys4mb372g__86mndsr0000gr/T/go-build375712479=/tmp/go-build -gno-record-gcc-switches -fno-common"
```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
INFO [config_reader] Config search paths: [./ /Users/david/dev/go/goyacc-tutorial /Users/david/dev/go /Users/david/dev /Users/david /Users /] 
INFO [config_reader] Used config file .golangci.yaml 
INFO [lintersdb] Active 91 linters: [asasalint asciicheck bidichk bodyclose containedctx contextcheck decorder depguard dogsled dupl dupword durationcheck errcheck errchkjson errname errorlint execinquery exhaustive exportloopref forbidigo forcetypeassert gci ginkgolinter gocheckcompilerdirectives gochecknoglobals gochecknoinits gocognit goconst gocritic gocyclo godot godox goerr113 gofmt gofumpt goheader goimports gomoddirectives gomodguard goprintffuncname gosec gosimple gosmopolitan govet grouper importas ineffassign interfacebloat ireturn lll loggercheck maintidx makezero mirror misspell musttag nakedret nestif nilerr nilnil nlreturn noctx nolintlint nonamedreturns nosprintfhostport paralleltest prealloc predeclared promlinter reassign revive rowserrcheck sqlclosecheck staticcheck stylecheck tagalign tagliatelle tenv testableexamples thelper tparallel typecheck unconvert unparam unused usestdlibvars wastedassign whitespace wrapcheck wsl zerologlint] 
INFO [loader] Go packages loading at mode 575 (compiled_files|deps|files|name|exports_file|imports|types_sizes) took 124.363166ms 
INFO [runner/filename_unadjuster] Pre-built 2 adjustments in 679.75µs 
INFO [linters_context] importas settings found, but no aliases listed. List aliases under alias: key. 
INFO [linters_context/goanalysis] analyzers took 7.192892032s with top 10 stages: the_only_name: 746.243165ms, buildir: 726.599623ms, buildssa: 481.265414ms, musttag: 291.126125ms, bidichk: 172.961ms, exhaustive: 166.152752ms, gocritic: 139.751375ms, wastedassign: 122.412875ms, dupl: 117.625583ms, goimports: 105.126376ms 
WARN [runner] Can't run linter goanalysis_metalinter: whitespace: failed to get line /Users/david/dev/go/goyacc-tutorial/ast_calculator/yaccpar:329: failed to get file /Users/david/dev/go/goyacc-tutorial/ast_calculator/yaccpar lines cache: can't get file /Users/david/dev/go/goyacc-tutorial/ast_calculator/yaccpar bytes from cache: can't read file /Users/david/dev/go/goyacc-tutorial/ast_calculator/yaccpar: open /Users/david/dev/go/goyacc-tutorial/ast_calculator/yaccpar: no such file or directory 
INFO [runner] processing took 1.497µs with stages: max_same_issues: 375ns, skip_dirs: 250ns, identifier_marker: 166ns, cgo: 125ns, skip_files: 42ns, path_prettifier: 42ns, autogenerated_exclude: 42ns, filename_unadjuster: 42ns, max_from_linter: 42ns, sort_results: 42ns, path_prefixer: 42ns, exclude-rules: 41ns, uniq_by_line: 41ns, path_shortener: 41ns, severity-rules: 41ns, nolint: 41ns, max_per_file_from_linter: 41ns, diff: 41ns, source_code: 0s, fixer: 0s, exclude: 0s 
INFO [runner] linters took 1.126762458s with stages: goanalysis_metalinter: 1.126732375s 
ERRO Running error: 1 error occurred:
	* can't run linter goanalysis_metalinter: whitespace: failed to get line /Users/david/dev/go/goyacc-tutorial/ast_calculator/yaccpar:329: failed to get file /Users/david/dev/go/goyacc-tutorial/ast_calculator/yaccpar lines cache: can't get file /Users/david/dev/go/goyacc-tutorial/ast_calculator/yaccpar bytes from cache: can't read file /Users/david/dev/go/goyacc-tutorial/ast_calculator/yaccpar: open /Users/david/dev/go/goyacc-tutorial/ast_calculator/yaccpar: no such file or directory
 
INFO Memory: 14 samples, avg is 167.2MB, max is 281.7MB 
INFO Execution took 1.265224291s                  
```

</details>

### Code example or link to a public repository

<details>

```shell
git clone git@github.com:slrtbtfs/goyacc-tutorial.git
cd goyacc-tutorial
go mod init github.com/slrtbtfs/goyacc-tutorial
printf "linters:\n  enable-all: true\n" > .golangci.yaml
golangci-lint run -v ./calculator
perl -i -pe 's{^//line yacc.+}{}' calculator/y.go
golangci-lint run -v ./calculator
```

</details>

### Validation

- [X] Yes, I've included all information above (version, config, etc.).

**Comment by ldez** [comment]
Hello,

> - [X] Yes, I've tried with the standalone linter if available (e.g., gocritic, go vet, etc.). (https://golangci-lint.run/usage/linters/)

you have checked this box, can you give me the output of the [whitespace](https://github.com/ultraware/whitespace) linter?

**Comment by ldez** [comment]
This linter is an old linter that doesn't follow our current standard for a linter, so there is no way to run it as a standalone linter :cry:

**Comment by theory** [comment]
Thanks. I worked around this issue by using the `-l` option to `goyacc` to omit the `//line` lines.

**Comment by bombsimon** [comment]
The `go/parser` treats this the same way the compiler:

> The `//line` directive specifies that the source line that follows should be recorded as having come from the given file path and line number. Successive lines are recorded using increasing line numbers, until the next directive. This directive typically appears in machine-generated code, so that compilers and debuggers will show lines in the original input to the generator.

So if you ask for the `file.Position(pos).Filename` for a node that is from a file with a `//line` comment the returned value will be based on this.

This code will print `yaccpar`. Removing the comment will instead print `src.go`.

```go
package main

import (
	"fmt"
	"go/ast"
	"go/parser"
	"go/token"
)

const src = `package main

//line yaccpar:1
func main() {}`

func main() {
	fset := token.NewFileSet()
	file, _ := parser.ParseFile(fset, "src.go", src, parser.ParseComments)
	decl := file.Decls[0].(*ast.FuncDecl)
	fmt.Println(fset.Position(decl.Body.Pos()).Filename)
}
```

Thus this doesn't seem to be isolated to `whitespace` but any linter that would have a fixer where we would call `GetLine` from a `token.Pos` coming from a file/node with a `//line` comment.

Sine we pass each file to `whitespace.Run` I was hoping we could just keep track of the filename inside `golangci-lint` for later lookups but for some reason that doesn't seem to work. We no longer crash but the fixer doesn't return the expected result for the current test.

<details>
  <summary>My filename attempt</summary>
  
```diff
diff --git a/pkg/golinters/whitespace.go b/pkg/golinters/whitespace.go
index e5941fa5..0f509208 100644
--- a/pkg/golinters/whitespace.go
+++ b/pkg/golinters/whitespace.go
@@ -63,9 +63,19 @@ func NewWhitespace(settings *config.WhitespaceSettings) *goanalysis.Linter {
 }
 
 func runWhitespace(lintCtx *linter.Context, pass *analysis.Pass, wsSettings whitespace.Settings) ([]goanalysis.Issue, error) {
-	var messages []whitespace.Message
+	type whitespaceIssue struct {
+		whitespaceMessage whitespace.Message
+		filename          string
+	}
+
+	var messages []whitespaceIssue
 	for _, file := range pass.Files {
-		messages = append(messages, whitespace.Run(file, pass.Fset, wsSettings)...)
+		for _, message := range whitespace.Run(file, pass.Fset, wsSettings) {
+			messages = append(messages, whitespaceIssue{
+				whitespaceMessage: message,
+				filename:          pass.Fset.Position(file.Pos()).Filename,
+			})
+		}
 	}
 
 	if len(messages) == 0 {
@@ -73,10 +83,12 @@ func runWhitespace(lintCtx *linter.Context, pass *analysis.Pass, wsSettings whit
 	}
 
 	issues := make([]goanalysis.Issue, len(messages))
-	for k, i := range messages {
+	for k, wi := range messages {
+		i := wi.whitespaceMessage
+
 		issue := result.Issue{
 			Pos: token.Position{
-				Filename: i.Pos.Filename,
+				Filename: wi.filename,
 				Line:     i.Pos.Line,
 			},
 			LineRange:   &result.Range{From: i.Pos.Line, To: i.Pos.Line},
@@ -85,9 +97,9 @@ func runWhitespace(lintCtx *linter.Context, pass *analysis.Pass, wsSettings whit
 			Replacement: &result.Replacement{},
 		}
 
-		bracketLine, err := lintCtx.LineCache.GetLine(issue.Pos.Filename, issue.Pos.Line)
+		bracketLine, err := lintCtx.LineCache.GetLine(wi.filename, issue.Pos.Line)
 		if err != nil {
-			return nil, fmt.Errorf("failed to get line %s:%d: %w", issue.Pos.Filename, issue.Pos.Line, err)
+			return nil, fmt.Errorf("failed to get line %s:%d: %w", wi.filename, issue.Pos.Line, err)
 		}
 
 		switch i.Type { 
```
</details>

**Comment by ldez** [comment]
@bombsimon whitespace is unmaintained :wink:

**Comment by bombsimon** [comment]
Yeah I saw the thread but I also saw someone merged a PR in `funlen` [4 hours ago](https://github.com/ultraware/funlen/commit/f53ef7f18f4282fbbf0ede84ace38a3865630f2b) so it doesn't seem _completely_ dead. Maybe the change is simple enough to get a merge anyways. 🤞

**Comment by ldez** [comment]
Maybe my email had its effect...

https://github.com/ultraware/funlen/issues/15#issuecomment-1648025030

**Comment by robinknaapen** [comment]
@bombsimon

Yes, our apologies for the delays. We are striving to maintain the packages. As said in https://github.com/ultraware/funlen/issues/15#issuecomment-1648025030

We are open to pull request. And we will strive to discuss/review/maintain the repositores

**Comment by ldez** [comment]
so whitespace be written to follow our current standard (the custom code inside golangci-lint should be removed)
@bombsimon can you handle that?

**Comment by bombsimon** [comment]
> so whitespace should be written to follow our current standard (the custom code inside golangci-lint should be removed) @bombsimon can you handle that?

Yes, I can do that. I have limited time the following ~1-2 weeks but I'll start and open a new PR when ready.

**Comment by bombsimon** [comment]
@ldez Any suggestions on how to tackle the fixer if converting to analyzer and not returning any result to `golangci-lint`? Should I traverse the `SuggestedFixes` and use them before #1779 is fixed or do you have a better solution?

**Comment by ldez** [comment]
you can do something like that: https://github.com/4meepo/tagalign/blob/43dc5d806d929c9454a82a2116728b45e1b6f0f0/tagalign.go#L158-L191

**Comment by bombsimon** [comment]
This is now fixed in #4003
